app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,208 @@
+        serializer = RecipeSerializer(all_recipes, many=True)
+
+        response = self.client.get(RECIPES_URL)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data), 2)
+        self.assertEqual(response.data, serializer.data)
+
+    def test_get_all_recipes_with_name_filter(self):
+        sample_recipe()
+        recipe = sample_recipe({
+            'name': 'A chicken dinner',
+            'description': 'A clucking good meal'
+        })
+
+        serializer = RecipeSerializer(recipe)
+
+        filter_term = 'chicken'
+        response = self.client.get(
+            RECIPES_URL,
+            {'name': filter_term}
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertIn(serializer.data, response.data)
+
+    def test_get_all_recipes_with_name_filter_no_results(self):
+        """Test retrieving a list of recipes when filter returns no matches"""
+
+        sample_recipe({
+            'name': 'Bangers and Mash',
+            'description': 'A hearty meal'
+        })
+
+        filter_term = 'Chips'
+        response = self.client.get(
+            RECIPES_URL,
+            {'name': filter_term}
+        )
+
+        # Then
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data), 0)
+
+    def test_get_recipe_by_id_success(self):
+        """Test retrieving a recipe by unique ID"""
+
+        sample_recipe()
+        recipe = sample_recipe({
+            'name': 'Bangers and Mash',
+            'description': 'A hearty meal'
+        })
+        recipe.ingredients.add(sample_ingredient(recipe=recipe, name='Mash'))
+
+        serializer = RecipeSerializer(recipe)
+
+        url = recipe_url(recipe.id)
+        response = self.client.get(url)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data, serializer.data)
+
+    def test_get_recipe_by_id_not_found(self):
+        """Test retrieving a recipe by ID when the recipe doesn't exist"""
+
+        response = self.client.get(recipe_url(recipe_id=1))
+
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_create_recipe(self):
+        """Test creating a new recipe"""
+
+        ingredient_name_1 = 'Sausage'
+        ingredient_name_2 = 'Gravy'
+
+        payload = {
+            'name': 'Bangers And Mash',
+            'description': 'Whack it all together',
+            'ingredients': [
+                {'name': ingredient_name_1},
+                {'name': ingredient_name_2},
+            ]
+        }
+
+        response = self.client.post(RECIPES_URL, payload, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        recipe = Recipe.objects.get(id=response.data['id'])
+        ingredients = recipe.ingredients.all()
+
+        self.assertEqual(ingredients.count(), 2)
```

Comment:
Alternatively, with one query less
```suggestion
        number_of_ingredients = Ingredient.objects.filter(recipe__id=response.data["id"]).count()
        self.assertEqual(number_of_ingredients, 2)
```